pkg/planner/core/logical_plan_builder.go

```diff
@@ -4763,7 +4768,7 @@ func (b *PlanBuilder) buildDataSource(ctx context.Context, tn *ast.TableName, as
 	if dirty || tableInfo.TempTableType == model.TempTableLocal || tableInfo.TableCacheStatusType == model.TableCacheStatusEnable {
 		us := logicalop.LogicalUnionScan{HandleCols: handleCols}.Init(b.ctx, b.getSelectOffset())
 		us.SetChildren(ds)
-		if tableInfo.Partition != nil && b.optFlag&rule.FlagPartitionProcessor == 0 {
+		if tableInfo.Partition != nil && !hasGlobal {
 			// Adding ExtraPhysTblIDCol for UnionScan (transaction buffer handling)
 			// Not using old static prune mode
```

Comment:
maybe change the comments here.

Reply:
I have revert this changement. it has been fixed by https://github.com/pingcap/tidb/pull/61089.